openlibrary/core/ia.py

```diff
@@ -2,24 +2,252 @@
+        # We need this to be http, so that we see the urls being called in the logs
+        # internally for rate limiting
+        return 'http://%s/advancedsearch.php?%s' % (
+            h.bookreader_host() or 'archive.org', urlencode(params, doseq=True))
+
+    @classmethod
+    def _compose_browsable_url(cls, query, sorts=None):
+        """If the client wants to explore this query in the browser, it will
+        have to be converted to a human readable version of the advancedsearch API
+        :param str query:
+        :param list of str sorts:
+        :rtype: str
+        :return: a url for humans to view this query on archive.org
+        """
+        _sort = sorts[0] if sorts else ''
+        if ' desc' in _sort:
+            _sort = '-' + _sort.split(' ')[0]
+        elif ' asc' in _sort:
+            _sort = _sort.split(' ')[0]
+        return '%s/search.php?%s' % (
+            h.ia_domain(), urlencode({'query': query, 'sort': _sort}))
+
+    @classmethod
+    def _request(cls, url):
+        """
+        Hits archive.org advancedsearch.php API, returns matching items
+        :param str url:
+        :rtype: dict
+        :return: the `response` content of the advancesearch
+        """
+        try:
+            # TODO switch to requests
+            request = six.moves.urllib.request.Request(url)
+            # Internet Archive Elastic Search (which powers some of our
+            # carousel queries) needs Open Library to forward user IPs so
+            # we can attribute requests to end-users
+            client_ip = web.ctx.env.get('HTTP_X_FORWARDED_FOR', 'ol-internal')
+            request.add_header('x-client-id', client_ip)
+            response = six.moves.urllib.request.urlopen(
+                request, timeout=h.http_request_timeout()).read()
+            return simplejson.loads(response).get('response', {})
+        except Exception:  # TODO make more specific
+            logger.error("IAEditionSearch._request error", exc_info=True)
+            return []
+
+    @classmethod
+    def _add_availability_to_edition(cls, edition, item_index):
+        """
+        To avoid a 2nd network call to `lending.add_availability` reconstruct
+        availability info ad-hoc from archive.org advancedsearch results
+        :param openlibrary.core.models.Edition edition:
+        :param dict item_index: mapping ocaid -> metadata item
+        :rtype: Edition
+        :return: edition, modified to include availability
+        """
+        item = item_index[edition.ocaid]
+        edition.set_availability(
+            identifier=item.get('identifier', ''),
+            status=item.get(cls.AVAILABILITY_STATUS, 'error'),
+            num_loans=int(item.get('loans__status__num_loans', 0)),
+            num_waitlist=int(item.get('loans__status__num_waitlist', 0)),
+        )
+        return edition
+
+    @staticmethod
+    def _index_items_by_ocaid(items):
+        return dict((item['identifier'], item) for item in items)
```

Comment:
Why pay the function call overhead?
```suggestion
        return {item['identifier']: item for item in items}
```